packages/changed-elements-react/src/api/ChangesTreeDataProvider.ts

```diff
@@ -474,24 +474,42 @@ export class ChangesTreeDataProvider implements ITreeDataProvider {
       updateFunc,
     );
 
-    let targetNodes = await this._getChangedModelOrSubjectNodes(
+    const targetNodes = await this._getChangedModelOrSubjectNodes(
       targetIModel,
       onlyDeleted,
       true,
       updateFunc,
     );
 
     const currentModelIds = new Set(currentNodes.map((node: TreeNodeItem) => node.id));
-    targetNodes = targetNodes.filter((entry: TreeNodeItem) => !currentModelIds.has(entry.id));
-
-    this._models = [...currentNodes, ...targetNodes];
+    const uniqueNodes = targetNodes.filter((entry: TreeNodeItem) => !currentModelIds.has(entry.id));
+    const mergedNodes = this._findAndMergeModelChanges(currentNodes, targetNodes);
+    this._models = [...mergedNodes, ...uniqueNodes];
 
     // Release, as it will be unused after the creation of models
     this._rootEntries = undefined;
 
     return this._models;
   }
 
+  private _findAndMergeModelChanges = (currentNodes: ReadonlyArray<TreeNodeItem>, targetNodes: ReadonlyArray<TreeNodeItem>) => {
```

Comment:
Can you add documentation for this function? Also, could you check that this is really necessary after fixing the proper changeset id being opened for the iModel? The `_getModelChangesFromRootEntries` should have been putting `hasDeletions` to the model nodes appropriately.

I think that's what needs to be debugged to understand why the filter is not working on the deleted elements? Were those deleted elements not `rootEntries`? Causing them to not be looked at during that `_getModelChangesFromRootEntries` function?